shapely/strtree.py

```diff
@@ -213,24 +281,29 @@ def nearest(self, geom):
         >>> tree = STRtree ([Point(0, 0), Point(0, 0)])
         >>> tree.nearest(Point(0, 0)).wkt
         'POINT (0 0)'
+
         """
-        if self._n_geoms == 0:
+        if not self._initdata:
             return None
 
         envelope = geom.envelope
-
-        def callback(item1, item2, distance, userdata):
-            try:
-                geom1 = ctypes.cast(item1, ctypes.py_object).value
-                geom2 = ctypes.cast(item2, ctypes.py_object).value
-                dist = ctypes.cast(distance, ctypes.POINTER(ctypes.c_double))
-                lgeos.GEOSDistance(geom1._geom, geom2._geom, dist)
-                return 1
-            except Exception:
-                return 0
-
-        item = lgeos.GEOSSTRtree_nearest_generic(self._tree_handle, ctypes.py_object(geom), envelope._geom, \
-            lgeos.GEOSDistanceCallback(callback), None)
+        try:
+            geoms, values = zip(*self._initdata)
+        except TypeError:
+            geoms = self._initdata
+            values = self._initdata
+
+        @nearest_callback
+        def callback(value, geom, userdata):
+            value_geom = geoms[values.index(value)]
```

Comment:
When non-geometry objects are stored in the tree we need to look up the geometries they were inserted with.